### PR TITLE
ghc 8.10.5

### PR DESCRIPTION
**Pull request by fishtreesugar**
Created with `brew bump-formula-pr`.

`resource` blocks may require updates.

### CODE REVIEW

**Comment by carlocab** [comment]
I think we can try to build this for ARM: https://www.haskell.org/ghc/blog/20210309-apple-m1-story.html

**Comment by carlocab** [comment]
This part

https://github.com/Homebrew/homebrew-core/blob/28f77b7dd9e83f9ba674c516ea21d6601fc776ac/Formula/ghc.rb#L56-L57

probably needs to be changed to something like

https://github.com/Homebrew/homebrew-core/blob/28f77b7dd9e83f9ba674c516ea21d6601fc776ac/Formula/gmp.rb#L31-L34

**Comment by fishtreesugar** [comment]
thx  @carlocab. And there's a [aarch64-apple-darwin binary release](https://downloads.haskell.org/ghc/8.10.5/ghc-8.10.5-aarch64-apple-darwin.tar.xz) of 8.10.5, does it useful for bootstrapping?

**Comment by carlocab** [comment]
That is actually very useful. I didn't see it on the downloads page. Let's use that.

**Comment by carlocab** [comment]
```
<no location info>: error:
    Warning: Couldn't figure out LLVM version!
             Make sure you have installed LLVM between 10 and 12
```

**Comment by carlocab** [comment]
Mojave build failed with
```
Undefined symbols for architecture x86_64:
  "___darwin_check_fd_set_overflow", referenced from:
      _awaitEvent in libHSrts.a(Select.o)
ld: symbol(s) not found for architecture x86_64
```
Mojave SDK is probably too old. [1] This is a problem, given the number of formulae that we build with `ghc`.

[1] https://twitter.com/iferrthrowbrick/status/1247908035965726725?s=21

**Comment by carlocab** [comment]
ARM build fails while building the docs. I wonder if there's a way to skip those.

**Comment by carlocab** [comment]
It built on ARM 🎉 

Though, it looks like we need a bootstrap `cabal-install`. C.f https://github.com/haskell/cabal/issues/7433

**Comment by cho-m** [comment]
Intel Catalina/Big Sur:
```
Error: 5 failed steps!
brew linkage --test agda
brew linkage --test allureofthestars
brew test --retry --verbose haskell-language-server
brew test --retry --verbose youtube-dlc
brew test --retry --verbose youtube-dlc
```

**Comment by cho-m** [comment]
Nice that Mojave CI passed with older `ghc`.

The main problem is figuring out cabal on ARM.
May need to create a local release via bootstrap method.

**Comment by carlocab** [comment]
Yup. If you're interested, you can follow the instructions at https://github.com/haskell/cabal/issues/7433 to generate a bootstrap `cabal` we can upload somewhere and use for the build. You can probably use the pre-built GHC 8.10.5 that upstream provides to build it.

Otherwise, I'll try to find some time to log into an ARM node and generate one there.

File: Formula/ghc.rb [hunk 1]
```diff
@@ -17,7 +17,8 @@ class Ghc < Formula
   end
 
   depends_on "python@3.9" => :build
-  depends_on "sphinx-doc" => :build
+  depends_on "llvm" if Hardware::CPU.arm?
```
Comment (cho-m) [inline Formula/ghc.rb:20]: Currently, this causes `llvm` to be pulled in as a runtime dependency.

Comment (carlocab) [reply Formula/ghc.rb:20]: Pretty sure it should be. It should actually also still be using system Clang to compile.

**Comment by cho-m** [comment]
On side note, `ghc` doesn't like our `llvm` version number, though it doesn't actually impact anything. Just complains a lot on logs:
```
You are using an unsupported version of LLVM!
Currently only 10 to 12 is supported. System LLVM version: 12.0.0
We will try though...
```

**Comment by carlocab** [comment]
The real test will be if we can build something with GHC. We'll need a working `cabal-install` to work on that...

**Comment by elbamos** [comment]
I was able to bootstrap cabal.  
[cabal-install-3.5-arm64-darwin-11.4-bootstrapped.tar.xz.zip](https://github.com/Homebrew/homebrew-core/files/6612500/cabal-install-3.5-arm64-darwin-11.4-bootstrapped.tar.xz.zip)

**Comment by carlocab** [comment]
Can you share the JSON file you used to generate that?

**Comment by elbamos** [comment]
Sure, its the one I pulled from the latest comment at https://github.com/haskell/cabal/issues/7433

**Comment by angerman** [comment]
> ```
> <no location info>: error:
>     Warning: Couldn't figure out LLVM version!
>              Make sure you have installed LLVM between 10 and 12
> ```

This warning can be safely ignored. It has its origins around LLVM 2/3, when LLVM (and especially LLVMs IR) changed significantly enough that GHC needed to pin an LLVM version. And even today we risk having to debug LLVM issues across multiple versions. My take is that from a practical point of view the version restriction makes little sense as users might not have the tools or even ability to install the recommended LLVM. The recommended LLVM with 8.10 is LLVM9, which for aarch64-darwin isn't even an option.

So yea, just ignore the warning, and let's try to bump the accepted range in 8.10.6 to include 12.

**Comment by carlocab** [comment]
@angerman That's the warning from not having LLVM installed. The warning about the version is a bit more confusing:

> ```
> You are using an unsupported version of LLVM!
> Currently only 10 to 12 is supported. System LLVM version: 12.0.0
> We will try though...
> ```

The `Currently only 10 to 12 is supported` is a little ambiguous here.

**Comment by carlocab** [comment]
Fixing HLS in #78925.

**Review by carlocab — APPROVED** [review]
Merging this. Will fix `cabal-install` this afternoon. Thanks, @fishtreesugar.

**Comment by angerman** [comment]
> @angerman That's the warning from not having LLVM installed. The warning about the version is a bit more confusing:
> 
> > ```
> > You are using an unsupported version of LLVM!
> > Currently only 10 to 12 is supported. System LLVM version: 12.0.0
> > We will try though...
> > ```
> 
> The `Currently only 10 to 12 is supported` is a little ambiguous here.

They are both conceptually the same, and can both be safely ignored. I've added an upstream issue here: https://gitlab.haskell.org/ghc/ghc/-/issues/19967. Not quite sure why 12.0.0 is parsed as falling outside of [10,12].

**Comment by carlocab** [comment]
Can they both be ignored? I ignored the second one but not the first, since the first one preceded a build failure.

**Comment by Bo98** [comment]
> Mojave build failed with
> 
> ```
> Undefined symbols for architecture x86_64:
>   "___darwin_check_fd_set_overflow", referenced from:
>       _awaitEvent in libHSrts.a(Select.o)
> ld: symbol(s) not found for architecture x86_64
> ```
> 
> Mojave SDK is probably too old. [1] This is a problem, given the number of formulae that we build with `ghc`.
> 
> [1] https://twitter.com/iferrthrowbrick/status/1247908035965726725?s=21

This sounds like it could be an issue with the bootstrap binary. Does 8.10.4 work?

**Comment by carlocab** [comment]
> This sounds like it could be an issue with the bootstrap binary. Does 8.10.4 work?

To bootstrap with, you mean? I don't think so. The first commit here didn't update the bootstrap GHC, and that didn't build.

**Comment by Bo98** [comment]
Not with the same error though. If you are getting `___darwin_check_fd_set_overflow` errors then something is miscompiled and I suspect it is the prebuilts.

**Comment by carlocab** [comment]
Hm, ok. Let's try to bootstrap with 8.10.4 on Mojave. Will do that when I have time in a bit.

**Comment by elbamos** [comment]
Just trying to be helpful - is the binary I provided above not sufficient for bootstrapping purposes? I was attempting to provide what was proposed here https://github.com/Homebrew/homebrew-core/pull/78821#issuecomment-856205330 as sufficient.

If not, since I have this running on my M1 system, is there anything I can do to help you folks nail down the bootstrapped cabal?  

For reference, this when I did this, it was very straightforward.  All I had to do was copy down the json, export PATH, LDFLAGS, and CCFLAGS to point to homebrew's llvm, and the cabal bootstrapping script ran without incident.

**Pull request closed without merging**